crates/svm-runtime/src/import.rs

```diff
@@ -1,52 +1,173 @@
+use std::convert::{TryFrom, TryInto};
 use std::ffi::c_void;
-use std::ptr;
+use std::rc::Rc;
 
 use crate::Context;
-use svm_types::WasmType;
 
-use wasmer::{Export, FunctionType, Type as WasmerType};
-use wasmer_vm::{ExportFunction, VMContext, VMFunctionKind};
+use wasmer::{Export, Exportable, Function, FunctionType, RuntimeError, Store, Type, Val};
 
-#[derive(Debug, PartialEq, Clone)]
-pub struct Import {
-    pub name: String,
+use svm_ffi::{svm_byte_array, svm_env_t, svm_func_callback_t, svm_trap_t};
+use svm_types::{WasmType, WasmValue};
 
-    pub params: Vec<WasmType>,
+#[derive(Debug, Clone)]
+pub struct ExternImport {
+    name: String,
 
-    pub returns: Vec<WasmType>,
+    namespace: String,
 
-    pub func_ptr: *const c_void,
+    params: Vec<WasmType>,
+
+    returns: Rc<Vec<WasmType>>,
+
+    func: svm_func_callback_t,
+
+    host_env: *const c_void,
 }
 
-impl Import {
-    pub fn to_wasmer(&self, ctx: Context) -> Export {
+impl ExternImport {
+    pub fn new(
+        name: String,
+        namespace: String,
+        params: Vec<WasmType>,
+        returns: Vec<WasmType>,
+        func: svm_func_callback_t,
+        host_env: *const c_void,
+    ) -> Self {
+        Self {
+            name,
+            namespace,
+            params,
+            returns: Rc::new(returns),
+            func,
+            host_env,
+        }
+    }
+
+    pub fn wasmer_export(&self, store: &Store, ctx: &mut Context) -> (Export, *const svm_env_t) {
+        unsafe {
+            // The following code has been highly influenced by code here:
+            // https://github.com/wasmerio/wasmer/blob/7847acaae1e7a0eade13b65def1f3feeac95efd7/lib/c-api/src/wasm_c_api/externals/func.rs#L86
+
+            let returns = self.returns.clone();
+            let func = self.func;
+
+            let inner_callback =
+                move |env: &mut *mut svm_env_t, args: &[Val]| -> Result<Vec<Val>, RuntimeError> {
+                    let args: Vec<WasmValue> = wasmer_vals_to_wasm_vals(args)?;
+                    let args: svm_byte_array = args.into();
+
+                    let mut results = svm_ffi::alloc_wasm_values(&returns);
+                    let trap = func(*env, &args, &mut results);
+
+                    // manually releasing `args` internals
+                    args.destroy();
+
+                    if !trap.is_null() {
+                        let trap: Box<svm_trap_t> = Box::from_raw(trap);
+
+                        let err_msg: String = (*trap).into();
+                        let err = RuntimeError::new(err_msg);
+
+                        return Err(err);
+                    }
+
+                    match Vec::<WasmValue>::try_from(&results) {
+                        Ok(vals) => {
+                            // TODO: validate the returns types.
```

Comment:
In general, this should be done via the `host trampoline` as part of injecting the `results`.

I'm not sure we want to have this validation logic duplicated here

Reply:
Yes, can probably avoid doing it here.

Reply:
I've added validation of the `results` since SVM only allocates a zeroed-buffer before calling the host function.